lib/graphql/query/serial_execution/field_resolution.rb

```diff
@@ -32,7 +32,10 @@ def result
         # continue by "finishing" the value, eg. executing sub-fields or coercing values
         def get_finished_value(raw_value)
           raise raw_value if raw_value.instance_of?(GraphQL::ExecutionError)
-          return nil if raw_value.nil?
+          if raw_value.nil?
+            raise GraphQL::ExecutionError, "Type mismatch resolving '#{field.name}'" if field.type.kind.non_null?
```

Comment:
Right, I knew there was a reason I did it this way to begin with:
- Complex types (objects, lists, etc) don't define `coerce_result` at all, so non-null's implementation is dead *and* badly broken. I'll remove it.
- If we do the check in `NonNullResolution`, then we have to be able to call `field.type.resolve_type(nil)` for all types in order to *get* the resolver class, but this method isn't well-defined for interfaces and it's not obvious that it should be (see the fourth point in my original PR post).